lib/charms/zookeeper/v0/client.py

```diff
@@ -0,0 +1,210 @@
+        for host in self.hosts:
+            with ZooKeeperClient(host=host, client_port=client_port) as zk:
+                response = zk.srvr
+                if response.get("Mode") == "leader":
+                    self.leader = host
+
+        if not self.leader:
+            raise QuorumLeaderNotFoundError("quorum leader not found, probably not ready yet")
+
+    @property
+    def server_members(self) -> Set[str]:
+        """The current members within the ZooKeeper quorum.
+
+        Returns:
+            set: A set of ZK member strings
+                e.g {"server.1=10.141.78.207:2888:3888:participant;0.0.0.0:2181"}
+        """
+        with ZooKeeperClient(host=self.leader, client_port=self.client_port) as zk:
+            members, _ = zk.config
+
+        return set(members)
+
+    @property
+    def version(self) -> int:
+        """The current config version for ZooKeeper.
+
+        Returns:
+            int: the zookeeper config version
+        """
+        with ZooKeeperClient(host=self.leader, client_port=self.client_port) as zk:
+            _, version = zk.config
+
+        return version
+
+    @property
+    def members_syncing(self) -> bool:
+        """Flag to check if any quorum members are currently syncing data.
+
+        Returns:
+            bool: True if any members are syncing
+        """
+        with ZooKeeperClient(host=self.leader, client_port=self.client_port) as zk:
+            result = zk.mntr
+        if (
+            result.get("zk_peer_state", "") == "leading - broadcast"
+            and result["zk_pending_syncs"] == "0"
+        ):
+            return False
+        return True
+
+    def add_members(self, members: Iterable[str]) -> None:
+        """Adds new members to the members' dynamic config.
+
+        Raises:
+            MembersSyncingError: if any members are busy syncing data
+            MemberNotReadyError: if any members are not yet broadcasting
+        """
+        if self.members_syncing:
+            raise MembersSyncingError("Unable to add members - some members are syncing")
+
+        for member in members:
+            host = member.split("=")[1].split(":")[0]
+
+            # individual connections to each server
+            with ZooKeeperClient(host=host, client_port=self.client_port) as zk:
+                if not zk.is_ready:
+                    raise MemberNotReadyError(f"Server is not ready: {host}")
+
+            # specific connection to leader
+            with ZooKeeperClient(host=self.leader, client_port=self.client_port) as zk:
+                zk.client.reconfig(
+                    joining=member, leaving=None, new_members=None, from_config=self.version
+                )
+
+    def remove_members(self, members: Iterable[str]):
+        """Removes members from the members' dynamic config.
+
+        Raises:
+            MembersSyncingError: if any members are busy syncing data
+        """
+        if self.members_syncing:
```

Comment:
might be worth checking if another member is getting removed, I know this is something we need to check in MongoDB